DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,257 @@
+
+void main(){
+    auto letterToday = WordLetterOfTheDay.b;
+    
+    import std.stdio;
+    switch(letterToday){
+        case $a:
+            writeln("Apple");
+            break;
+        case $b:
+            writeln("Bicycle");
+            break;
+        case $c:
+            writeln("Caterpillar");
+            break;
+        case $d:
+            writeln("Didgeridoo");
+            break;
+        /*...*/
+    }
+}
+```
+
+### 5. Array literals
+A) ETI is allowed in array literals in initializations and assignments as determined by item 1 (Initializations and assignments) above.
+
+B) ETI is also allowed in array literals for which an explicit enum type can be inferred.
+```d
+enum A{ a,b,c,d }
+
+// (A)
+A[4] x = [$a, $b, $c, $d];
+// (B)
+auto y = [A.a, $b, $c, $d]; // typeof(y) = A[]
+auto z = [A.c, 64, $b, $b]; // typeof(z) = int[]
+```
+
+### Other considerations
+When the resolution of ETI is ambiguous due to multiple candidate enums, ETI is prohibited.
+```d
+enum A{ a,b,c,d,e }
+enum B{ a,b,c,d }
+
+void myFn(A param){}
+void myFn(B param){}
+
+void main(){
+    myFn(A.a);
+    myFn($a); // error, we have two equally valid candidates!
+    myFn($e); // OK, the only candidate is "myFn(A.e)", since B does not have "$e"
+}
+```
+
+### ETI vs `with` statements
+For some use-cases, an argument could be made against the need for ETI. For instance, `final switch` statements combined with `with` statements:
+```d
+enum A{ a, b, c }
+
+void main(){
+    auto myA = A.b;
+    final switch(myA) with(typeof(myA)){
+        case a:
+            // ...
+            break;
+        case b:
+            // ...
+            break;
+        case c:
+            // ...
+            break;
+    }
+}
+```
+In the above case, ETI is not necessary.
+
+However, there are many everyday cases where using `with` statements would prove insufficient, or outright silly.
+For example, given the following delcarations:
+```d
+enum Size{ small, medium, large }
+enum Shape{ square, round, triangular }
+enum State{ on, off, undefined }
+
+struct Obj{
+    Size  size;
+    Shape shape;
+    State state;
+    
+    void myFn(Size size, Shape shape, State state){};
+}
+```
+Let's create a `main` that doesn't use ETI or `with` statements:
+```d
+void main(){
+    Obj myObj = {
+        size:  Size.large,
+        shape: Shape.round,
+        state: State.on,
+    };
+    myObj.myFn(Size.medium, Shape.square, State.undefined);
+}
+```
+Next, here's a version using ETI:
+```d
+void main(){
+    Obj myObj = {
+        size:  $large,
+        shape: $round,
+        state: $on,
+    };
+    myObj.myFn($medium, $square, $undefined);
+}
+```
+And finally, here's a version rewritten to use `with` statements:
+```d
+void main(){
+    Obj myObj;
+    with(myObj){
+        with(Size)  size  = large;
+        with(Shape) shape = round;
+        with(State) state = on;
+        with(Size){
+            myFn(medium, Shape.square, State.undefined);
+        }
+    }
+}
```

Comment:
`with` statemets can do better than that:
```D
void main()
{   Obj myObj;
    with(myObj) with(Size) with(Shape) with(State){
        size  = large;
        shape = round;
        state = on;

        myFn(medium, square, undefined);
    }
}
```

Now granted, a style checking tool (or a human style police of a rigid sort) might not accept that. But I argue that's a problem of the style, not of the language.

Reply:
That's cursed, but also even less readable than my original example. What's large? Is it a variable, a function, a field or member of myObj, a member of Size, a member of Shape, a member of State... etc. Too many things in scope, brain melted!

Reply:
P.S. yeah I'll update the example to match that